crates/pallet-domains/src/runtime_registry.rs

```diff
@@ -122,43 +131,78 @@ pub(crate) fn register_runtime_at_genesis<T: Config>(
+pub(crate) fn do_schedule_runtime_upgrade<T: Config>(
     runtime_id: RuntimeId,
     code: Vec<u8>,
-    at: T::BlockNumber,
-) -> Result<(), Error> {
-    RuntimeRegistry::<T>::try_mutate(runtime_id, |maybe_runtime_object| {
-        let runtime_obj = maybe_runtime_object
-            .as_mut()
-            .ok_or(Error::MissingRuntimeObject)?;
-
-        let new_runtime_version = can_upgrade_code(&runtime_obj.version, &code)?;
-        let runtime_hash = T::Hashing::hash(&code);
-
-        runtime_obj.code = code;
-        runtime_obj.version = new_runtime_version;
-        runtime_obj.hash = runtime_hash;
-        runtime_obj.runtime_upgrades = runtime_obj
-            .runtime_upgrades
-            .checked_add(1)
-            .ok_or(Error::MaxRuntimeUpgrades)?;
-        runtime_obj.updated_at = at;
-        Ok(())
-    })
+    current_block_number: T::BlockNumber,
+) -> Result<T::BlockNumber, Error> {
+    let runtime_obj = RuntimeRegistry::<T>::get(runtime_id).ok_or(Error::MissingRuntimeObject)?;
+    let new_runtime_version = can_upgrade_code(&runtime_obj.version, &code)?;
+    let scheduled_at = current_block_number
+        .checked_add(&T::DomainRuntimeUpgradeDelay::get())
+        .ok_or(Error::MaxScheduledBlockNumber)?;
+    let scheduled_upgrade = ScheduledRuntimeUpgrade {
+        code,
+        version: new_runtime_version,
+    };
+    ScheduledRuntimeUpgrades::<T>::insert(scheduled_at, runtime_id, scheduled_upgrade);
+    Ok(scheduled_at)
+}
+
+#[derive(Default)]
+pub(crate) struct UpgradeRuntimesResult {
+    pub(crate) reads: u64,
+    pub(crate) writes: u64,
+}
+
+pub(crate) fn do_upgrade_runtimes<T: Config>(at: T::BlockNumber) -> UpgradeRuntimesResult {
+    let mut upgrade_runtimes_result = UpgradeRuntimesResult::default();
+
+    for (runtime_id, scheduled_update) in ScheduledRuntimeUpgrades::<T>::drain_prefix(at) {
+        RuntimeRegistry::<T>::mutate(runtime_id, |maybe_runtime_object| {
+            let runtime_obj = maybe_runtime_object
+                .as_mut()
+                .expect("Runtime object exists since an upgrade is scheduled after verification");
+
+            let runtime_hash = T::Hashing::hash(&scheduled_update.code);
+            runtime_obj.code = scheduled_update.code;
+            runtime_obj.version = scheduled_update.version;
+            runtime_obj.hash = runtime_hash;
+            runtime_obj.runtime_upgrades = runtime_obj.runtime_upgrades.saturating_add(1);
+            runtime_obj.updated_at = at;
+        });
+
+        // one read for each key under `ScheduledRuntimeUpgrades`
+        upgrade_runtimes_result.reads += 1;
+        // two writes, one to kill the `ScheduledRuntimeUpgrades` and one to update `RuntimeRegistry`
+        upgrade_runtimes_result.writes += 2;
+
+        // deposit digest log for light clients
+        frame_system::Pallet::<T>::deposit_log(DigestItem::domain_runtime_upgrade(runtime_id));
```

Comment:
Actually, we don't have to worry about the weights at all right now, just leave a TODO and we can add the benchmark properly later. The manual weight is pretty error-prone, e.g., the weight of the og deposit is missing.

Reply:
> The manual weight is pretty error-prone, e.g., the weight of the log deposit is missing.

Manual record is required here due to the number of the runtime that might get upgrade. Happy to leave a todo to reveist but will leave the current chnages as is though

Reply:
As I said, the current UpgradeRuntimesResult is incorrect, depositing a log causes 1 read and 1 write, which is missing, so better not to introduce such code from the beginning, just leave a TODO and calculate the weight on the fly precisely later using something `benchmark_upgrade_one_runtime() * number_of_upgraded_runtimes`.

Reply:
I have adjusted and left a todo. Thanks!

Reply:
System Events is typically whitelisted from the benchmarking, so the adjusted version still produces a different result from the proper benchmark. I'd say just removing everything about the manual weight is the simplest for now.